### DISCUSSION TITLE
Rewrites not working in combination with i18n

### DISCUSSION THREAD

**Discussion by dennispassway** [vercel/next.js#19707] [Help]
Not sure if this is supposed to be here or at the Vercel repo. Sorry if it's in the wrong place. We have an issue with rewrites in combination with using next's new i18n. We're using next 10.0.3.

We have the following file structure:
```
- pages
  * index.js
  * [slug].js
- public
  * robots.txt
  * og-image.png
  - cms
    * index.html
```

And the configuration in next.config.js containing the rewrites:
```javascript
{
  ...,
  async rewrites() {
    return [{ source: "/cms/(.*)", destination: "/cms" }];

  },
  i18n: {
    locales: ["en", "nl"],
    defaultLocale: "nl",
  },
}
```

When I visit `/cms` I get a 404. When visiting `/cms/` (trailing slash), I correctly get rewritten to the CMS single page application. The CMS SPA redirects (client side) to `/cms/dashboard`, which works. But on refresh, it returns to 404. So the `/cms/[something]` page is not rewritten.

I have also tried the following rewrites pattern, with the same result.
```javascript
{
  ...,
  async rewrites() {
    return [
      { source: "/cms", destination: "/cms/index.html", locale: false },
      { source: "/cms/", destination: "/cms/index.html", locale: false },
      { source: "/cms/:path*", destination: "/cms/index.html", locale: false },
    ];
  },
  i18n: {
    locales: ["en", "nl"],
    defaultLocale: "nl",
  },
}
```

We have not changed anything relevant between releases and have found that our build started breaking 7 days ago, on November 24th. A change I found in Vercel that might be relevant is https://github.com/vercel/vercel/commit/fee386493b32e16a7f9227db8d6cb229c6ba8d57. I'm not sure if that causes our issue but I wonder if something changed in Vercel that makes our rewrites break.

I believe I have set up our rewrites as stated in the docs, but I cannot find an example for rewrites in combination with i18n and the `/public` directory. Did I set it up wrong is there something changed in Vercel or Next causing this issue?

**Comment by dennispassway** [comment]
This is still an issue for us. If I update nextjs to 10.0.3 we get an issue in Vercel stating: "An unexpected error happened when running this build. We have been notified of the problem.". In 10.0.1 we dont have this issue.

**Reply by Aratramba** [reply]
We're still seeing this in 10.0.5

**Comment by ijjk** [comment]
In the above `locale: false` rewrites you need to specify the locale in each `source` manually as mentioned in the docs here https://nextjs.org/docs/api-reference/next.config.js/rewrites#rewrites-with-i18n-support or they won't be matched. 

Note: the source `/cms/(.*)` is saying to only match `/cms/` or `/cms/some-path` not `/cms`, breaking the rewrites out into `/cms` and `/cms/(.*)` does correct this but it appears `locale: false` should not be used here since you want the locale to be handled automatically.

**Comment by Aratramba** [comment]
thanks @ijjk for your reponse. We've resolved the problem. I don't think this is an issue with Next. 

I'll post the solution here for future reference.

next.config.js:
```
module.exports = {
  async rewrites() {
    return [
      { source: "/cms", destination: "/cms/index.html" },
      { source: "/cms/:path*", destination: "/cms/index.html" },
    ];
  },
  serverRuntimeConfig: {
    rootDir: __dirname,
  },
  i18n: {
    locales: ["en", "nl"],
    defaultLocale: "nl",
  },
  basePath: "",
};
```

We also had a now.json in our project which caused the problem.
```
{
  "version": 2,
  "cleanUrls": false
}
```

I think the `cleanUrls: false` setting (https://vercel.com/docs/configuration#project/clean-urls) causes a circular rewrite between `/cms/index.html` getting a 308 redirect to `/cms` which gets rewritten to `/cms/index.html`. Removing this setting, fixed the issue. 

I'm not sure why this returns a 404 and not a server error.

The broken deployments on Vercel are caused by returning `{ notFound: true }` in our pages and have nothing to do with the issue at hand. We've notified the Vercel team of this issue.

**Comment by winsvold** [comment]
I have a similar case, but we also have a basePath. Made a new discussion here, would be grateful for any help or insights on this :) https://github.com/vercel/next.js/discussions/22449
